ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/operators/KeyedTableLookupOperator.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.physical.pull.operators;
+
+import com.google.common.collect.ImmutableList;
+import io.confluent.ksql.execution.streams.materialization.Locator.KsqlPartitionLocation;
+import io.confluent.ksql.execution.streams.materialization.Materialization;
+import io.confluent.ksql.execution.streams.materialization.Row;
+import io.confluent.ksql.planner.plan.DataSourceNode;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import org.apache.kafka.connect.data.Struct;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class KeyedTableLookupOperator
+    extends AbstractPhysicalOperator
+    implements UnaryPhysicalOperator, DataSourceOperator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(KeyedTableLookupOperator.class);
+
+  private final Materialization mat;
+  private final DataSourceNode logicalOperator;
+
+  private List<KsqlPartitionLocation> partitionLocations;
+  private Iterator<Row> resultIterator;
+
+  public KeyedTableLookupOperator(
+      final Materialization mat,
+      final DataSourceNode logicalNode
+  ) {
+    this.logicalOperator = Objects.requireNonNull(logicalNode, "logicalNode");
+    this.mat = Objects.requireNonNull(mat, "mat");
+  }
+
+  @Override
+  public void open() {
+    final List<Row> result = new ArrayList<>();
+    for (KsqlPartitionLocation location : partitionLocations) {
+      if (!location.getKeys().isPresent()) {
+        throw new IllegalStateException("Table lookup queries should be done with keys");
+      }
+      for (Struct key : location.getKeys().get()) {
```

Comment:
It's slightly funny to do the real work in open and then just dish out the cached rows on each next.  Since the interface is row oriented, we could also keep track of the key we're on and then do the actual single key lookup on each call to next.  In theory, this allows for "back pressure" and avoiding a burst a work since we could do it incrementally and on demand.

Reply:
There are multiple locations and keys, so we need to keep track of both in next and exhaust both iterators. I made the change, lmk how it looks

Reply:
I first commented and then waited for the tests to pass before pushing. Unfortunately, it took longer than expected :(